Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h

```diff
@@ -656,6 +658,69 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
     return JSValue::encode(thisObject);
 }
 
+template<typename ViewClass>
+ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncFromFast(VM& vm, JSGlobalObject* globalObject, CallFrame* callFrame)
+{
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue arrayLike = callFrame->uncheckedArgument(1);
+    JSArrayBufferView* items = jsDynamicCast<JSArrayBufferView*>(arrayLike);
+    if (!items) {
+        // TypedArray.from(Array) case.
+        JSArray* array = jsDynamicCast<JSArray*>(arrayLike);
+        if (!array)
+            return JSValue::encode(jsUndefined());
+
+        if (!array->isIteratorProtocolFastAndNonObservable())
+            return JSValue::encode(jsUndefined());
+
+        IndexingType indexingType = array->indexingType() & IndexingShapeMask;
+        if (indexingType != Int32Shape && indexingType != DoubleShape)
+            return JSValue::encode(jsUndefined());
+
+        size_t length = array->length();
+
+        Structure* structure = globalObject->typedArrayStructure(ViewClass::TypedArrayStorageType);
+        ViewClass* result = ViewClass::createUninitialized(globalObject, structure, length);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (indexingType == Int32Shape) {
+            for (unsigned i = 0; i < length; i++) {
+                JSValue value = array->butterfly()->contiguous().at(array, i).get();
+                if (LIKELY(!!value))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromInt32(value.asInt32()));
+                else
+                    result->setIndexQuickly(i, jsUndefined());
```

Comment:
Is this valid? From what I can tell, `setIndexQuickly` can only handle a JSValue that is int32 or double.  See `Adaptor::Type toNativeFromValue(JSValue value)`.

Reply:
Oops, it should be `toNativeFromValue(globalObject, ...` conversion.

Reply:
We already know the source came from an Int32Shape contiguous array, and in this case, we have a hole.  Why go through a function to convert `undefined` to a value?  Isn't it always 0?

Reply:
If the array is Int32Shape and iterator protocol is ensured, then we are ensured that we will get an undefined value here.
But what value will be stored into `result` depends on `ViewClass`. For example, if it is `Int32Array`, `0` will be stored. But if it is `Float64Array`, then it should get `NaN` since undefined will be converted to `NaN` when converting it to number.
So, that's why we need `toNativeFromValue`. It is ensured that this behavior is unobservable since it will not invoke any functions. Converting `undefined` to `ViewClass`'s type. But still only way we have is `toNativeFromValue`.
I wonder if we should add `toNativeFromUndefined`, which would be useful in the other place too. Trying.

Reply:
Ah, wait, it is fun. BigInt64Array can get an error if `undefined` is converted! So, I think we should skip this optimization path for BigInt64Array etc. But it is fine since it cannot accept these Int32Shape etc. anyway

Reply:
I'll fix it soon.

Reply:
Ah, and it just works with `toNativeFromDouble(PNaN)`.